iceoryx_posh/source/roudi/service_registry.cpp

```diff
@@ -17,45 +17,153 @@
+            }
+            it++;
+        }
+    };
+
+    removeIndexFromMap(m_serviceMap, index);
+    removeIndexFromMap(m_instanceMap, index);
+
+    return removedElement;
 }
 
-void ServiceRegistry::find(InstanceSet_t& instances,
+void ServiceRegistry::find(ServiceDescriptionVector_t& searchResult,
                            const capro::IdString_t& service,
-                           const capro::IdString_t& instance) const
+                           const capro::IdString_t& instance) const noexcept
 {
-    if (instance == Wildcard)
+    cxx::vector<uint64_t, MAX_SERVICE_DESCRIPTIONS> intersection;
+
+    // Find (K1, K2)
+    // O(log n + log n + max(#PossibleServices + #possiblesInstances) + #intersection)
+    if (instance != Wildcard && service != Wildcard)
     {
-        for (auto& instance : m_serviceMap[service].instanceSet)
+        cxx::vector<uint64_t, MAX_SERVICE_DESCRIPTIONS> possibleServices;
+        cxx::vector<uint64_t, MAX_SERVICE_DESCRIPTIONS> possibleInstances;
+
+        auto rangeServiceMap = m_serviceMap.equal_range(service);
+        for (auto entry = rangeServiceMap.first; entry != rangeServiceMap.second; ++entry)
+        {
+            possibleServices.push_back(entry->second);
+        }
+
+        auto rangeInstanceMap = m_instanceMap.equal_range(instance);
+        for (auto entry = rangeInstanceMap.first; entry != rangeInstanceMap.second; ++entry)
+        {
+            possibleInstances.push_back(entry->second);
+        }
+
+        ::std::set_intersection(possibleServices.begin(),
+                                possibleServices.end(),
+                                possibleInstances.begin(),
+                                possibleInstances.end(),
+                                ::std::back_inserter(intersection));
+
+        for (auto& value : intersection)
         {
-            instances.push_back(instance);
+            searchResult.push_back(m_serviceDescriptionVector[value]);
         }
     }
-    else
+    // Find (*, K2)
+    // O(log n + #result)
+    else if (service == Wildcard && instance != Wildcard)
+    {
+        auto range = m_instanceMap.equal_range(instance);
+        for (auto entry = range.first; entry != range.second; ++entry)
+        {
+            searchResult.push_back(m_serviceDescriptionVector[entry->second]);
+        }
+    }
+    // Find (K1, *)
+    // O(log n + #result)
+    else if (instance == Wildcard && service != Wildcard)
     {
-        auto& instanceSet = m_serviceMap[service].instanceSet;
-        auto iter = std::find(instanceSet.begin(), instanceSet.end(), instance);
-        if (iter != instanceSet.end())
+        auto range = m_serviceMap.equal_range(service);
+        for (auto entry = range.first; entry != range.second; ++entry)
         {
-            instances.push_back(*iter);
+            searchResult.push_back(m_serviceDescriptionVector[entry->second]);
         }
     }
+    else
+    {
+        // Find (*, *)
+        // O(1)
```

Comment:
No, it is `O(n)` if we copy the whole vector since `#result = n` here. We measure the effort as function of `n`, the number of elements which it certainly depends on.

This is why passing only a reference to the underlying container (if possible) would gain a lot of efficiency, since then it would be `O(1)`. I suppose the operation is only used rarely and we can think about this later.